src/com/amazon/corretto/crypto/provider/EcParameters.java

```diff
@@ -0,0 +1,115 @@
+package com.amazon.corretto.crypto.provider;
+
+import java.io.IOException;
+import java.lang.reflect.Method;
+import java.security.*;
+import java.security.spec.*;
+
+
+public final class EcParameters extends AlgorithmParametersSpi {
+    private EcUtils.ECInfo ecInfo;
+
+    // A public constructor is required by AlgorithmParameters class.
+    public EcParameters() {}
+
+    protected void engineInit(AlgorithmParameterSpec paramSpec) throws InvalidParameterSpecException {
+        if (paramSpec == null) {
+            throw new InvalidParameterSpecException("paramSpec must not be null");
+        }
+
+        String name = null;
+        if (paramSpec instanceof ECParameterSpec) {
+            name = EcUtils.getNameBySpec((ECParameterSpec) paramSpec);
+            // Earlier Java TLS implementations cache curve params by OID instead of human-readable "shortname".
+            // Specifying the shortname where OID is expected results in TLS handshake failures due to the server
+            // not being able to "find" the intended curve, tricking the server into believing it doesn't support any
+            // of the curves stipulated by the client.
+            //
+            // https://github.com/corretto/corretto-8/blob/235873fd43e5b7aa556d011f436e65a99c10c20a/jdk/src/share/classes/sun/security/ssl/SupportedGroupsExtension.java#L585-L607
+            // https://github.com/openjdk/jdk10u/blob/ef9178d7d8a4489640a31a1d0c88958724af5304/src/java.base/share/classes/sun/security/ssl/SupportedGroupsExtension.java#L188-L210
+            if (Utils.getJavaVersion() < 11) {
+                name = EcUtils.getOidFromName(name);
+            }
+        } else if (paramSpec instanceof ECGenParameterSpec) {
+            name = ((ECGenParameterSpec) paramSpec).getName();
+        } else if ("sun.security.util.ECKeySizeParameterSpec".equals(paramSpec.getClass().getName())) {
+            // OpenJDK's JCE sometimes passes ECKeySizeParameterSpec when initializing EC AlgorithmParameters. We can
+            // reference this class directly using instanceof up until Java 8, but in Java 9 it was moved to the
+            // java.base module and only exported to other JDK-internal modules. To work around this, we use reflection
+            // to get the key size and get the corresponding curve name from our own database. Future versions of java
+            // may restrict reflective access to private modules, so this functionality may break.
+            //
+            // https://github.com/corretto/corretto-11/blob/14c02261590b4dc01284888a7a51d39ff581ac8d/src/java.base/share/classes/sun/security/util/ECParameters.java#L121
+            // https://github.com/corretto/corretto-11/blob/14c02261590b4dc01284888a7a51d39ff581ac8d/src/java.base/share/classes/module-info.java#L301-L314
+            try {
+                Method getKeySize = paramSpec.getClass().getMethod("getKeySize");
+                Integer keySize = Integer.class.cast(getKeySize.invoke(paramSpec));
```

Comment:
Standard reflection due diligence - We assume this interface is rock solid at this point since this is only necessary for older JDKs?

Reply:
very fair. in my testing, it's actually necessary for all JDKs tested under integration tests (8, 11, 17). the difference with 8 is that we _could_ import it there and test with `instanceof` instead of using reflection. in later versions, reflection is necessary due to modularization in JDK9, but not always possible. it's not possible unless `--add-opens=java.base/sun.security.util=ALL-UNNAMED` or similar is specified in JVM invocation.

this reflective call is designed to try the reflective check, and if it is not possible to fall back "cleanly" to the exception case on L57.

ECKeySizeParameterSpec is a JDK-internal class, so i wouldn't really call it an "interface", but the goal here is to fail in a "clean" way when the required permissions aren't available.

all this said, i think we need a test case for the negative case when reflective permissions aren't allowed. i'll work on adding that.